kolibri/core/content/utils/content_manifest.py

```diff
@@ -0,0 +1,195 @@
+        return self._channels_dict.keys()
+
+    def get_channel_versions(self, channel_id):
+        return self._channels_dict.get(channel_id, {}).keys()
+
+    def get_channel_data(self, channel_id, channel_version):
+        channel_data = self._channels_dict.get(channel_id, {}).get(
+            channel_version, None
+        )
+        if channel_data is None:
+            channel_data = ContentManifestChannelData(None, None, set())
+        return channel_data
+
+    def get_include_node_ids(self, channel_id, channel_version):
+        channel_data = self.get_channel_data(channel_id, channel_version)
+        return channel_data.include_node_ids
+
+    def add_content_nodes(self, channel_id, channel_version, nodes_queries_list):
+        include_node_ids = get_content_nodes_selectors(
+            channel_id, channel_version, nodes_queries_list
+        )
+        self._update_channel_data(channel_id, channel_version, include_node_ids)
+
+    def _update_channel_data(self, channel_id, channel_version, include_node_ids):
+        old_include_node_ids = self.get_include_node_ids(channel_id, channel_version)
+        channel_data = ContentManifestChannelData(
+            channel_id,
+            channel_version,
+            set(old_include_node_ids) | set(include_node_ids),
+        )
+        self._set_channel_data(channel_data)
+
+    def _set_channel_data(self, channel_data):
+        assert isinstance(channel_data, ContentManifestChannelData)
+        self._channels_dict.setdefault(channel_data.channel_id, {})[
+            channel_data.channel_version
+        ] = channel_data
+
+
+def get_content_nodes_selectors(channel_id, channel_version, nodes_queries_list):
+    """
+    Returns a set of include_node_ids that can be used with the given
+    channel_id to import only the nodes contained in nodes_queries_list.
+    """
+
+    include_node_ids = set()
+
+    channel_metadata = ChannelMetadata.objects.get(
+        id=channel_id, version=channel_version
+    )
+
+    available_node_ids = set(
+        itertools.chain.from_iterable(
+            nodes_query.values_list("id", flat=True)
+            for nodes_query in nodes_queries_list
+        )
+    )
+
+    available_nodes_queue = [channel_metadata.root]
+
+    while len(available_nodes_queue) > 0:
+        node = available_nodes_queue.pop(0)
+
+        # We could add nodes to exclude_node_ids when less than half of the
+        # sibling nodes are missing. However, it is unclear if this would
+        # be useful.
+
+        if node.kind == "topic":
+            leaf_node_ids = _get_leaf_node_ids(node)
+            matching_leaf_nodes = leaf_node_ids.intersection(available_node_ids)
+            missing_leaf_nodes = leaf_node_ids.difference(available_node_ids)
+            if len(missing_leaf_nodes) == 0:
+                include_node_ids.add(node.id)
+            elif len(matching_leaf_nodes) > 0:
+                available_nodes_queue.extend(node.children.all())
```

Comment:
Should we be optionally filtering the node children by an `available` flag here (similarly to how we do in the import export data function).

Reply:
Oh, yeah, I think that's right. Assuming the availability of nodes in the tree is correct, this should be doing `node.children.filter(available=True)`. Unavailable nodes aren't going to resolve to anything useful here so it's just wasting memory. I should make sure the tests actually cover this properly first (might need to adopt that channel builder instead of the questionable database fixture :b), so, I'll do that on Monday.

I see what you mean about passing an `available=` flag to the function to be consistent with the import export data function, although I should tweak some names in that case, and probably wrap my head about the use case there. It does _sound_ like it would be convenient for something.

Reply:
Yeah, looking at how the function is used, I'm not sure there is a need to have optional available filtering. It's really just that it can be filtered by available=True to limit the search space.

The only question in my mind is.... does the manifest handle the case where a ContentNode has the files it needs in the exported directory, but is not available on the machine doing the exporting?

So, for example, if I exported the basic arithmetic topic from Khan Academy from my machine, and it generated a content manifest - I then pass that USB key to you, and you export more of Khan Academy from your machine, but you don't have basic arithmetic on yours - is that going to now be excluded from the updated manifest?

Would be good to make a test case to cover this scenario, I think.

Reply:
Okay, I added some test cases which explore that. Specifically:

- If we add some new content nodes to a ContentManifest, `include_node_ids` is _extended_ to include those nodes. Any nodes that were listed previously are still there.
- If content nodes are added for one channel version, and then for a different version of the same channel, the two versions are separated; the lists of `include_node_ids` are not merged. (In practice, importcontent merges them in #9467, albeit with a warning message).

This behaviour means a bit of duplication in some cases. That can be solved, but it adds some complexity - especially if we start thinking about channel versions - so I think just something to be aware of for the moment.